astroquery/tests/test_cache.py

```diff
@@ -0,0 +1,218 @@
+    assert cache_conf.cache_active == default_active
+
+
+def test_basic_caching():
+    cache_conf.reset()
+
+    mytest = CacheTestClass()
+    assert cache_conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    set_response(TEXT1)
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT1
+    assert len(os.listdir(mytest.cache_location)) == 1
+
+    set_response(TEXT2)
+
+    resp = mytest.test_func(URL2)  # query that has not been cached
+    assert resp.content == TEXT2
+    assert len(os.listdir(mytest.cache_location)) == 2
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT1  # query that was cached
+    assert len(os.listdir(mytest.cache_location)) == 2  # no new cache file
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    resp = mytest.test_func(URL1)
+    assert resp.content == TEXT2  # Now get new response
+
+
+def test_change_location(tmp_path):
+    cache_conf.reset()
+
+    mytest = CacheTestClass()
+    default_cache_location = mytest.cache_location
+
+    assert paths.get_cache_dir() in str(default_cache_location)
+    assert "astroquery" in mytest.cache_location.parts
+    assert mytest.name in mytest.cache_location.parts
+
+    new_loc = tmp_path.joinpath("new_dir")
+    mytest.cache_location = new_loc
+    assert mytest.cache_location == new_loc
+
+    mytest.reset_cache_location()
+    assert mytest.cache_location == default_cache_location
+
+    new_loc.mkdir(parents=True, exist_ok=True)
+    with paths.set_temp_cache(new_loc):
+        assert str(new_loc) in str(mytest.cache_location)
+        assert "astroquery" in mytest.cache_location.parts
+        assert mytest.name in mytest.cache_location.parts
+
+
+def test_login():
+    cache_conf.reset()
+
+    mytest = CacheTestClass()
+    assert cache_conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    set_response(TEXT1)  # Text 1 is set as the approved password
+
+    mytest.login("ceb")
+    assert mytest.authenticated()
+    assert len(os.listdir(mytest.cache_location)) == 0  # request should not be cached
+
+    set_response(TEXT2)  # Text 2 is not the approved password
+
+    mytest.login("ceb")
+    assert not mytest.authenticated()  # Should not be accessing cache
+
+
+def test_timeout(monkeypatch):
+    cache_conf.reset()
+
+    mytest = CacheTestClass()
+    assert cache_conf.cache_active
+
+    mytest.clear_cache()
+    assert len(os.listdir(mytest.cache_location)) == 0
+
+    set_response(TEXT1)  # setting the response
+
+    resp = mytest.test_func(URL1)  # should be cached
+    assert resp.content == TEXT1
+
+    set_response(TEXT2)  # changing the response
+
+    resp = mytest.test_func(URL1)  # should access cached value
+    assert resp.content == TEXT1
+
+    # Changing the file date so the cache will consider it expired
+    cache_file = list(mytest.cache_location.iterdir())[0]
```

Comment:
The preferred way of obtaining a value from a generator is by using `next`:
```suggestion
    cache_file = next(mytest.cache_location.iterdir())
```
But this is a minor remark.